api/types/container/network.go

```diff
@@ -0,0 +1,338 @@
+package container
+
+import (
+	"errors"
+	"fmt"
+	"iter"
+	"strconv"
+	"strings"
+	"unique"
+)
+
+// NetworkProtocol represents a network protocol for a port.
+type NetworkProtocol string
+
+const (
+	TCP  NetworkProtocol = "tcp"
+	UDP  NetworkProtocol = "udp"
+	SCTP NetworkProtocol = "sctp"
+)
+
+// Sentinel port proto value for zero Port and PortRange values.
+var protoZero unique.Handle[NetworkProtocol]
+
+// Port is a type representing a single port number and protocol in the format "<portnum>/[<proto>]".
+//
+// The zero port value, i.e. Port{}, is invalid; use [ParsePort] to create a valid Port value.
+type Port struct {
+	num   uint16
+	proto unique.Handle[NetworkProtocol]
+}
+
+// ParsePort parses s as a [Port].
+//
+// It normalizes the provided protocol such that "80/tcp", "80/TCP", and "80/tCp" are equivalent.
+// If a port number is provided, but no protocol, the default ("tcp") protocol is returned.
+func ParsePort(s string) (Port, error) {
+	if s == "" {
+		return Port{}, errors.New("value is empty")
+	}
+
+	port, proto, _ := strings.Cut(s, "/")
+
+	portNum, err := parsePortNumber(port)
+	if err != nil {
+		return Port{}, fmt.Errorf("invalid port '%s': %w", port, err)
+	}
+
+	normalizedPortProto := normalizePortProto(proto)
+	return Port{num: portNum, proto: normalizedPortProto}, nil
+}
+
+// MustParsePort calls [ParsePort](s) and panics on error.
+//
+// It is intended for use in tests with hard-coded strings.
+func MustParsePort(s string) Port {
+	p, err := ParsePort(s)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// PortFrom returns a [Port] with the given number and protocol.
+//
+// If no protocol is specified (i.e. proto == ""), then PortFrom returns Port{}, false.
+func PortFrom(num uint16, proto NetworkProtocol) (p Port, ok bool) {
+	if proto == "" {
+		return Port{}, false
+	}
+	normalized := normalizePortProto(string(proto))
+	return Port{num: num, proto: normalized}, true
+}
+
+// Num returns p's port number.
+func (p Port) Num() uint16 {
```

Comment:
Slightly related to the above (also depend on how it's used).


(Also still hate that `net.JoinHostPort` requires a _string_ for the port 😠 😂 )